include/plugin-apis/impl-in-plugin/trace-abi.h

```diff
@@ -0,0 +1,56 @@
+#ifndef TRACE_ABI_H
```

Comment:
Probably I am abusing the term. What I am trying to convey is the idea that these headers are intended to describe what is needed in order to ensure binary compatibility, because I believe that binary compatibility offers more flexibility than compatibility at the level of C language/syntax (in particular, implementations do not have to be written in C or C++).

Concretely, this means that to the greatest extent possible, the important information conveyed these header files is not preprocessor definitions, macros, or typedefs, but instead things that map directly to memory layout and calling convention characteristics which are well-defined in the context of a particular host platform.

Even more concretely, when you write `typedef struct my_struct_t my_struct_t;`, I am saying that that does nothing to describe an ABI because it only says the syntax that you can use (you can write `my_struct_t` instead of `struct my_struct_t`). But when you write `typedef struct my_struct_t { int a; double b; }`, I am saying that that helps to describe an ABI because you are saying that you are interested in something that, on the current host platform, has the memory layout of a C struct with an `int` followed by a `double`. Similarly, the macros that allow logging to be excluded from compilation depending on the log level do not describe ABI, but the declarations of the logging functions do, as does the declaration of the function that, during initialization, checks that the core and the plugins are in agreement about the log level.

I also think that it is sometimes possible to have an abstraction layer that works at the level of the C language (API) but not at the binary level (what I have been calling ABI), AND that it is possible in some cases to fix that. For example, the platform abstraction uses typedefs to say what a mutex is, but this abstraction does not work at the binary level because at the binary level, it is observable and problematic that the `sizeof` the two kinds of mutexes is different. A way to solve this in a way that achieves binary compatibility (what I have been calling ABI) is to change the typedef to describe a POINTER to a mutex instead of the mutex itself. This way, the sizeof the different pointers to mutexes is all the same, and this results in a consistent binary interface because all of the functions that access the mutex can use these pointers while abstracting the fact that they point to things that have different memory layout.

Reply:
Ok, I think I understand where you are coming from and it is great to see that you are thinking about those aspects. I would say that "ABI" and "binary compatibility" are largely misleading here. True binary compatibility you can only achieve when using the same compiler, on the same OS, on the same architecture. The binary interface that actually gets generated is defined by the compiler based on the ABI of the system.

Let me share my view on what you are trying to do. You are trying to separate a generic interface that can be implemented in many different ways from those functions/tools/types/macros that provide a convenience layer on top. In the case of the logging API, for instance, there are the signatures of the log functions doing the actual work, and there is a convenience layer on top provided by the macros. The requirements for designing a user-facing API and an API that acts as an interface between exchangeable components are very different, and this surfaces in the argument you make. Also both are different artifacts, and it is a good idea to separate them. We can also consider giving these two APIs different names in our terminology. Maybe programmer interface and module interface?

> I also think that it is sometimes possible to have an abstraction layer that works at the level of the C language (API) but not at the binary level (what I have been calling ABI), AND that it is possible in some cases to fix that. For example, the platform abstraction uses typedefs to say what a mutex is, but this abstraction does not work at the binary level because at the binary level, it is observable and problematic that the sizeof the two kinds of mutexes is different. A way to solve this in a way that achieves binary compatibility (what I have been calling ABI) is to change the typedef to describe a POINTER to a mutex instead of the mutex itself. This way, the sizeof the different pointers to mutexes is all the same, and this results in a consistent binary interface because all of the functions that access the mutex can use these pointers while abstracting the fact that they point to things that have different memory layout.

I don't quite follow this specific argument. Let's say there is the core runtime and an extension, both of which use the platform abstraction. The extension will only be compatible with the core runtime if both are compiled using the same platform abstraction (for the same platform). If both are complied for the same platform, both agree about the precise type used and hence `sizeof` returns the same value both in the extension and in the core runtime. In a sense, if you expose a type that depends on compiler settings (i.e. macro definitions) on the module interface, then your interface depends on this setting and you can generate many different interfaces. In the case of the platform, I think this is what we want because we statically select the platform. 

By introducing a pointer indirection, the problem becomes masked but not solved. There is still an implicit contract because the core runtime is compiled using platform X with a specific type A and the extension is compiled using platform Y with a specific type B. If `A != B`, then the core and the extension agree on the size of the pointer type (assuming that we compile for the same ISA), but the program will expose undefined behavior because the implementations don't agree on the size and layout of the object. You would need polymorphism to solve this. Surely, this can be done, but why resolve types dynamically if we are statically compiling for a specific platform anyway? 

The pointer solution that you mention pared with a vtable for polymorphism provides a good module interface when you don't know in advance (at compile time) which type(s) you will be dealing with. But (hopefully) we will not have to deal with unknown platforms at runtime. Whenever we can avoid that indirection, we should.

Reply:
> ...we will not have to deal with unknown platforms at runtime

This is not about trying to deal with scenarios where at compile time we do not yet know the platform that the code will have to work with at runtime. That is not my motivation here.

> If both are complied for the same platform, both agree about the precise type used

...unless they are compiled using different toolchains, which might not even have any shared concept of what a "type" means. Sometimes I try to avoid explicitly using the word "Rust" because I would feel a little silly evangelizing any one particular toolchain, but maybe this pointlessly obscuring my point. Really this is all about being able to implement these interfaces using languages other than C. Being able to write code in a language other than C is my real motivation.

> but the program will expose undefined behavior because the implementations don't agree on the size and layout of the object.

But not if the extension never uses the object directly, but instead only ever passes around the pointer -- as would have to be the case if the layout of the object is scrubbed from the interface, and the extension implementor doesn't do anything devious. The functions that work on the object directly would be provided by the core, and the compilation of the extension would not, even under the hood, have to be parameterized by platform-specific knowledge encoded in the platform abstraction.

> The pointer solution that you mention paired with a vtable for polymorphism provides a good module interface when you don't know in advance (at compile time) which type(s) you will be dealing with

Yes, I guess it is accurate to compare what I am talking about (functions being provided by the core) as being like a vtable. I have been thinking more along the lines of these functions being provided at link time rather than their being provided using a vtable of function pointers, but in the case of a third-party plugin providing these functions at link time may indeed have performance consequences similar to the consequences of a vtable (this is why I talk about LTO in the "tracing plugin" internal RFC).

Note that I have not been aware of performance as a priority in the C target for some time, and I/we have not even set up continuous benchmarking in CI for this reason. So although it would be unreasonable for me to categorically dismiss performance discussions (including concerns about vtable pointer lookup overhead), it is true that any such discussions would be limited by not having the context of a shared holistic view of the actual performance behavior and desired performance targets of the C runtime.

Later you also write

> I don't think avoiding C syntax is a good goal.

We should definitely make sure we agree about what our goals are because if the interface is defined syntactically rather than as "the interface is whatever the calling convention/struct layout would be for this C code on the target platform", then that is not compatible with implementing plugins in languages other than C or maybe C++.

Reply:
My vtable argument was not about performance (but its try that this is an aspect, too), its about complexity. To me this additional layer seems a bit too much, but it also seems like I am still missing something. I wasn't considering the Rust binding before, that of course changes the problem a bit. I think I am still not fully understanding the problem though. Perhaps we should have a call next week and sort this out. It seems important for me to understand those underlying problems.

On the naming: I still think that ABI is misleading as ABIs come into play below the language level and are not at all concerned with types. It seems like what you are after is a simplified interface that is safe to be used between different languages. Not sure how to call this though. Maybe FFI? But that could be misleading too

Reply:
Shall we just carve out time for this during our weekly LF meeting?

Reply:
Sure, that would work good for me. If anyone prefers a smaller meeting with Erling and Christian, that's good for me too. Whatever works.

Reply:
I think we have reached a collective understanding of the situation. The term "ABI" has been removed from this PR because I agree that it does not capture the intended constraint that I would like to put on certain interfaces between modules.

Marking this thread as resolved.